### PR TITLE
raftstore: fix the recovery process hanging issue caused orphaned unsafe recovery state

### LINKED ISSUE #12644: [Online Recovery] Unsafe online recovery failed with info “unsafe recovery enters exit force leader stage exceeds timeout”
### What did you do?
1、down two tikv
2、execute online unsafe recovery

### What did you expect to see?
Unsafe recovery success

### What did you see instead?
  },
  {
    "info": "Unsafe recovery enters exit force leader stage",
    "time": "2022-05-25 09:22:54.728",
    "details": [
      "Exceeds timeout 2022-05-25 09:22:54.399620114 +0000 UTC m=+82128.320209995"
    ]
  },
  {
    "info": "Collecting reports from alive stores(0/2)",
    "time": "2022-05-25 09:22:54.728",
    "details": [
      "Stores that have not dispatched plan: 5136724, 5136696",
      "Stores that have reported to PD: ",
      "Stores that have not reported to PD: "
    ]

### What version of PD are you using (`pd-server -V`)?
/ # ./pd-server -V
Release Version: v6.1.0
Edition: Community
Git Commit Hash: a6689e0efec71d9b27c6ea45b1f9dd0e0ee11af5
Git Branch: heads/refs/tags/v6.1.0
UTC Build Time:  2022-05-23 01:59:27

### PR DESCRIPTION
**Pull request by v01dstar**
Signed-off-by: v01dstar <yang.zhang@pingcap.com>

### What is changed and how it works?

Issue Number: Close #12644

What's Changed:

```commit-message
Cleaning up the unsafe recovery state after exiting previous joint state before proposing the recovery demotion which may return early if any error happens and leave the state untouched. 
```

### Related changes

- PR that adds the recovery logic and introduces the bug: #12022
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

- Performance regression
N/A
- Breaking backward compatibility
N/A

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -1999,15 +1999,16 @@ where
             }) => {
                 if self.fsm.peer.raft_group.raft.raft_log.applied >= *target_index {
                     if *demote_after_exit {
+                        let syncer_clone = syncer.clone();
+                        let failed_voters_clone = failed_voters.clone();
+                        self.fsm.peer.unsafe_recovery_state = None;
                         if !self.fsm.peer.is_force_leader() {
                             error!(
                                 "Unsafe recovery, lost forced leadership after exiting joint state";
                                 "region_id" => self.region().get_id(),
```
Comment (Connor1996) [inline components/raftstore/src/store/fsm/peer.rs:2008]: add peer_id

Comment (Connor1996) [reply components/raftstore/src/store/fsm/peer.rs:2008]: and other places

Comment (v01dstar) [reply components/raftstore/src/store/fsm/peer.rs:2010]: Added

**Review by Connor1996 — APPROVED** [review]
LGTM

**Review by 5kbpers — APPROVED** [review]

**Pull request merged**